src/constraints.jl

```diff
@@ -727,3 +727,313 @@ function list_of_constraint_types(model::Model)
+            end
+            push!(basis_status, con_basis_status)
+        end
+    end
+    return basis_status
+end
+
+function perturbation_range_of_feasibility(constraint::ConstraintRef{Model, _MOICON{F, S}}
+                      ) where {S <: Union{MOI.LessThan, MOI.GreaterThan, MOI.EqualTo}, T, F <: Union{MOI.ScalarAffineFunction{T}, MOI.SingleVariable}}
+    model = constraint.model
+    if termination_status(model) != MOI.OPTIMAL
+        error("The range of feasibility is not available because the current solution "*
+              "is not optimal.")
+    end
+
+    # This throw the error: ... do not support accessing ... ConstraintBasisStatus(), if its not implemented by the solver.
+    con_status = MOI.get(model, MOI.ConstraintBasisStatus(), constraint)
+    # The constraint is inactive.
+    if con_status == MOI.BASIC
+        if S <: MOI.LessThan
+            return (value(constraint) - constraint_object(constraint).set.upper, Inf)
+        end
+        if S <: MOI.GreaterThan
+            return (- Inf, value(constraint) - constraint_object(constraint).set.lower)
+        end
+        return (0.0, 0.0)
+    end
+    # The constraint is active.
+
+    # Optimization possible, only basic columns needed
+    A, L, U, constraints = _std_matrix(model)
+    var_basis_status = _std_basis_status(model)
+    x = _std_primal_solution(model, constraints)
+    # Ax = 0 => B*x_B = -N x_N =: b_N
+    # Perturb row i by Δ => ̃x_B = B^-1 (b_N + Δ e_i) = x_B + Δ ρ
+    # Perturb var bound j by Δ => ̃x_B = B^-1 (b_N - Δ N_j) = x_B + Δ ρ
+    # Find min and max values of Δ : L_B <= ̃x_B <= U_B
+    # L_B <= x_B + Δ ρ <= U_B  -> L_B - x_B <= Δ ρ <= U_B - x_B
+    # Δ <= (U_B - x_B)_+ ./ ρ_+,  (L_B - x_B)_- ./ p_-
+    # Δ >= (U_B - x_B)_- ./ ρ_-,  (L_B - x_B)_+ ./ p_+
+    basic = [var_basis_status[i] == MOI.BASIC for i in 1:length(var_basis_status)] # .== throws ERROR: MethodError: no method matching length(::MathOptInterface.BasisStatusCode)
+    B = A[:, basic]
+    x_B = x[basic]
+    local rho::Vector{Float64}
+    if F <: MOI.SingleVariable
+        var = constraint_object(constraint).func
+        j = findfirst(all_variables(model) .== var)
+        rho = -B \ Vector(A[:, j])
+    else
+        i = findfirst(constraints .== constraint)
+        ei = zeros(size(A)[1])
+        ei[i] = 1.0
+        rho = B \ ei
+    end
+    UmX_B = U[basic] - x_B
+    LmX_B = L[basic] - x_B
+    pos = rho .> 1e-7
+    neg = rho .< -1e-7
+    return (maximum([-Inf; UmX_B[neg] ./ rho[neg]; LmX_B[pos] ./ rho[pos]]),
+            minimum([Inf; UmX_B[pos] ./ rho[pos]; LmX_B[neg] ./ rho[neg]]))
+end
+
+"""
+Returns the optimal reduced costs for the variables of the problem in standard form, c.f. `_std_matrix(model)`
+"""
+function _std_reduced_costs(model::Model, constraints::Vector{ConstraintRef})
+    con_types = list_of_constraint_types(model)
+    vars = all_variables(model)
+    var2idx = Dict{VariableRef, Int64}()
```

Comment:
The [JuMP style guide](https://www.juliaopt.org/JuMP.jl/latest/style/#Abbreviations-1) would prefer `variable_to_column`